### DISCUSSION TITLE
Web app support

### DISCUSSION THREAD

**Discussion by skeledrew** [beeware/toga#1322] [Ideas]
I am evaluating Beeware for an app, and I would like for it to have a web version. However I saw a note that web support should be considered experimental. I am wondering at what stage this experimentation is, if there is already some at least somewhat framework in place, particularly for Toga. Having read up on its architecture, it seems it might not be too much work to use the [Brython project](https://brython.info) for example as the bridge to Javascript and web components at the native level. As I am also evaluating Brython, I would not be averse to potentially making contributions along these lines (if I proceed with these projects).

Thoughts? Is adding web support more involved than I am considering it to be?

**Comment by freakboy3742** [comment]
Toga Web should be considered highly experimental at present. We had a working demo a couple of years back, but that demo (and the capabilities that fed it) haven't been maintained.

That said, if you're keen to contribute, we're keen for any assistance you want to provide.

Regarding Brython specifically: I have some reservations over whether Brython is the right approach for the long term - I think WASM offers a much more viable alternative. However, the task of executing Python code in the client is almost orthogonal to the task of serving the web page that has that content; so I can see a world where a "Brython engine" is just one deployment option for the Python execution component that could one day be replaced by a native WASM approach. 

There's also the argument that a working but problematic something is better than a theoretically better but non existent nothing - and having *something* that proves the concept would potentially increase interest in developing the WASM alternative backend.

As for whether it would be a difficult project? Well, it won't be *trivial* - but as with most Toga backends, once the core of the problem is solved, implementing individual widgets is mostly labor intensive, rather than complex. If you can get Tutorial 0 running (a button that pops up a dialog when you press it), you've solved the two biggest problems: how do you render an app using Toga's layout algorithm, and how do you run Python code in response to user events. 

Web does have one additional concern - working out how to map URLs space onto navigation - but that's a design question that can be tackled once the short term questions about widgets have been resolved.

**Reply by skeledrew** [reply]
Was the reason it wasn't maintained due to priorities? As in would it be "fairly" straight-forward to carry on from where things stopped? Or is there a primary architectural, etc concern that needs reworking and would likely trigger a rewrite? How can I test this demo (if it still works)?

I also think WASM is a huge part of the future (and have been somewhat following the [Pyodide project](https://pyodide.org/en/stable/)). However, according to the [site FAQ](https://webassembly.org/docs/faq/), WASM is intended to complement instead of replace JS and so I think it will take a while yet for WASM to gain feature parity with JS, if ever. That being said, given that Toga is architected to support multiple backends, I also see the case for Brython/WASM/other as optional engines targeting the same platform. But yes, it would be very nice to have at least some fledgling support that others needing it could help build out.

I don't have much experience building for the web and so doing something like this would also be a huge learning experience. I have implemented a few examples with Brython and have found it pretty straight forward thus far. I will take a look at some of the Toga code and see how I can build on it.

**Reply by freakboy3742** [reply]
Yes - the lack of progress is mostly due to priorities (or, at least, priorities combined with the amount of free time available for contribution). 

The last time the web demo actually worked was documented in [this conference talk](https://www.youtube.com/watch?v=xezYYBL7nk0). It's been a while (almost 4 years... 😓 ) since that demo, and Toga has undergone a fairly major internal change since that code ran - that demo was on the 0.2 branch, and we've been in 0.3 development since mid 2018. I've done some *very* preliminary porting of the web backend to 0.3, but honestly, you'll be pretty much starting from scratch.

If you want to know more about where I think WASM fits into the picture, here are a couple of talks I've given on the topic:
* https://www.youtube.com/watch?v=2XSeNQyPlTY
* https://www.youtube.com/watch?v=M7KDhPG7vHw

If you're keen to look into this, then by all means take a look. As I said in my last response - the hard part is getting "hello world" going; once you've got that, the rest should be *reasonably* straight forward. It will definitely be a good learning experience - if you've got the patience to bang your head against the hard problems for a while :-) - but we're here to provide pointers if you need them.
